File: source/lsp_user_actions.cpp
/**
    Copyright (C) 2023 - 2026, edgo authors

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, see <https://www.gnu.org/licenses/>.
*/
#include "lsp_model.h"
#include "lsp_client.h"

hoverresponse_t Hover(std::string file, const int line, const int character) {
	lspclient.id++;

	auto request = baserequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/hover",
		params: params_t {
			textDocument: textdocument_t { uri: "file://" + file },
			position: position_t { line: line, character: character },
		},
	};

	send<baserequest_t>(request);
	auto response = WaitForRequest<hoverresponse_t>(&lspclient.userMessages, 1000);

	return response;
}

completionresponse_t Completion(std::string file, int line, int character) {
	lspclient.id++;

	auto request = baserequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/completion",
		params: params_t {
			textDocument: textdocument_t { uri:  "file://" + file },
			position: position_t { line: line, character: character },
			context: context_t { triggerKind: 1 },
		},
	};

	send<baserequest_t>(request);
	auto response = WaitForRequest<completionresponse_t>(&lspclient.userMessages, 1000);

	return response;
}

definitionresponse_t Definition(std::string file, int line, int character) {
	lspclient.id++;

	auto request = definitionrequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/definition",
		params: definitionparams_t {
			textDocument: textdocument_t { uri: "file://" + file },
			position: position_t { line: line, character: character },
		},
	};

	send<definitionrequest_t>(request);
	auto response = WaitForRequest<definitionresponse_t>(&lspclient.userMessages, 1000);

	return response;
}

signaturehelpresponse_t SignatureHelp(std::string file, int line, int character) {
	lspclient.id++;

	auto request = baserequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/signatureHelp",
		params: params_t {
			textDocument: textdocument_t { uri: "file://" + file },
			position: position_t { line: line, character: character },
		},
	};

	send<baserequest_t>(request);
	auto response = WaitForRequest<signaturehelpresponse_t>(&lspclient.userMessages, 1000);

	return response;
}

referencesresponse_t References(std::string file, int line, int character) {
	lspclient.id++;

	auto request = baserequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/references",
		params: params_t {
			textDocument: textdocument_t { uri: "file://" + file },
			position: position_t { line: line, character: character },
			context: context_t { includeDeclaration: false },
		},
	};

	send<baserequest_t>(request);
	auto response = WaitForRequest<referencesresponse_t>(&lspclient.userMessages, 3000);

	return response;
}


preparerenameresponse_t PrepareRename(std::string file, int line, int character) {
	lspclient.id++;

	auto request = preparerenamerequest_t {
		id: lspclient.id, jsonrpc: "2.0", method:  "textDocument/prepareRename",
		params: params_t {
			textDocument: textdocument_t { uri:  "file://" + file },
			position: position_t { line: line, character: character },
		},
	};

	send<preparerenamerequest_t>(request);
	auto response = WaitForRequest<preparerenameresponse_t>(&lspclient.userMessages, 10000);

	return response;
}

renameresponse_t Rename(std::string file, std::string newname, int line, int character) {
	lspclient.id++;

	auto request = renamerequest_t {
		id: lspclient.id,  jsonrpc: "2.0", method:  "textDocument/rename",
		params: renameparams_t {
			newName: newname,
			position: position_t { line: line, character: character },
			textDocument: textdocument_t { uri:  "file://" + file },
		},
	};

	send<renamerequest_t>(request);
	auto response = WaitForRequest<renameresponse_t>(&lspclient.userMessages, 10000);

	return response;
}

codeactionresponse_t CodeAction(std::string file, int spc, int spl, int epc, int epl) {
	lspclient.id++;

	auto request = codeactionrequest_t {
		id: lspclient.id, jsonrpc: "2.0", method: "textDocument/codeAction",
		params: codeactionparams_t {
			textDocument: textdocument_t { uri:  "file://" + file },
			context: context_t { only: string_v{"refactor"}, triggerKind: 1 },
			range: requestrange_t {
				start: position_t { line: spl, character: spc},
				end: position_t { line: epl, character: epc},
			},
		},
	};

	send<codeactionrequest_t>(request);
	auto response = WaitForRequest<codeactionresponse_t>(&lspclient.userMessages, 10000);

	return response;
}

File: legacy/lsp/lsp_user_actions.go
package lsp

func (this *LspClient) Hover(file string, line int, character int) (HoverResponse, error) {
	this.id++

	request := BaseRequest{
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/hover",
		Params: Params {
			TextDocument: TextDocument { URI: "file://" + file },
			Position: Position { Line: line, Character: character },
		},
	}

	this.send(request)
	response, err := WaitForRequest[HoverResponse](this.userMessages, 1000)

	return response, err
}

func (this *LspClient) Completion(file string, line int, character int) (CompletionResponse, error) {
	this.id++

	request := BaseRequest{
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/completion",
		Params: Params{
			TextDocument: TextDocument { URI:  "file://" + file },
			Position: Position { Line: line, Character: character },
			Context: Context { TriggerKind: 1 },
		},
	}

	this.send(request)
	response, err := WaitForRequest[CompletionResponse](this.userMessages, 1000)

	return response, err
}

func (this *LspClient) Definition(file string, line int, character int) (DefinitionResponse, error) {
	this.id++

	request := DefinitionRequest{
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/definition",
		Params: DefinitionParams {
			TextDocument: TextDocument{ URI: "file://" + file },
			Position: Position{ Line: line, Character: character },
		},
	}

	this.send(request)
	response, err := WaitForRequest[DefinitionResponse](this.userMessages, 1000)

	return response, err
}

func (this *LspClient) SignatureHelp(file string, line int, character int) (SignatureHelpResponse, error) {
	this.id++

	request := BaseRequest{
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/signatureHelp",
		Params: Params {
			TextDocument: TextDocument { URI: "file://" + file },
			Position: Position { Line: line, Character: character },
		},
	}

	this.send(request)
	response, err := WaitForRequest[SignatureHelpResponse](this.userMessages, 1000)

	return response, err
}

func (this *LspClient) References(file string, line int, character int) (ReferencesResponse, error) {
	this.id++

	request := BaseRequest{
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/references",
		Params: Params{
			TextDocument: TextDocument{ URI: "file://" + file },
			Position: Position{ Line: line, Character: character },
			Context: Context{ IncludeDeclaration: false },
		},
	}

	this.send(request)
	response, err := WaitForRequest[ReferencesResponse](this.userMessages, 3000)

	return response, err
}


func (this *LspClient) PrepareRename(file string, line int, character int) (PrepareRenameResponse, error) {
	this.id++

	request := PrepareRenameRequest {
		ID: this.id, JSONRPC: "2.0", Method:  "textDocument/prepareRename",
		Params: Params{
			TextDocument: TextDocument { URI:  "file://" + file },
			Position: Position { Line: line, Character: character },
		},
	}

	this.send(request)
	response, err := WaitForRequest[PrepareRenameResponse](this.userMessages, 10000)

	return response, err
}

func (this *LspClient) Rename(file string, newname string, line int, character int) (RenameResponse, error) {
	this.id++

	request := RenameRequest{
		ID: this.id,  JSONRPC: "2.0", Method:  "textDocument/rename",
		Params: RenameParams {
			NewName: newname,
			Position: Position { Line: line, Character: character },
			TextDocument: TextDocument { URI:  "file://" + file },
		},
	}

	this.send(request)
	response, err := WaitForRequest[RenameResponse](this.userMessages, 10000)

	return response, err
}

func (this *LspClient) CodeAction(file string, spc int, spl int, epc int, epl int) (CodeActionResponse, error) {
	this.id++

	request := CodeActionRequest {
		ID: this.id,  JSONRPC: "2.0", Method: "textDocument/codeAction",
		Params: CodeActionParams {
			TextDocument: TextDocument { URI:  "file://" + file },
			Context: Context{ Only: []string{"refactor"}, TriggerKind: 1 },
			Range: RequestRange{
				Start: Position{ Line: spl, Character: spc},
				End: Position{ Line: epl, Character: epc},
			},
		},
	}

	this.send(request)
	response, err := WaitForRequest[CodeActionResponse](this.userMessages, 10000)

	return response, err
}
